Tests/UpstreamIntegrationTests/util/AnyCodableUtils.swift

```diff
@@ -0,0 +1,666 @@
+import XCTest
+
+extension AnyCodable {
+    /// Converts `AnyCodable`'s default decode strategy of array `[Any?]`  into `[AnyCodable]` value type
+    public static func from(array: [Any?]?) -> [AnyCodable]? {
+        guard let unwrappedArray = array else { return nil }
+        
+        var newArray: [AnyCodable] = []
+        for val in unwrappedArray {
+            if let anyCodableVal = val as? AnyCodable {
+                newArray.append(anyCodableVal)
+            } else {
+                newArray.append(AnyCodable(val))
+            }
+        }
+        
+        return newArray
+    }
+    
+}
+
+class AnyCodableUtils {
+    // MARK: AnyCodable helpers
+    /// Performs testing assertions between two `[AnyCodable]` instances.
+    @discardableResult
+    static func assertEqual(lhs: [String: AnyCodable]?, rhs: [String: AnyCodable]?, keyPath: [Any], file: StaticString = #file, line: UInt = #line, shouldAssert: Bool = true) -> Bool {
+        if lhs == nil, rhs == nil {
+            return true
+        }
+        guard let lhs = lhs, let rhs = rhs else {
+            if shouldAssert {
+                XCTFail(#"""
+                    \#(lhs == nil ? "lhs is nil" : "rhs is nil") and \#(lhs == nil ? "rhs" : "lhs") is non-nil
+                
+                    lhs: \#(String(describing: lhs))
+                    
+                    rhs: \#(String(describing: rhs))
+                    
+                    key path: \#(keyPathAsString(keyPath: keyPath))
+                """#, file: file, line: line)
+            }
+            return false
+        }
+        if lhs.count != rhs.count {
+            if shouldAssert {
+                XCTFail(#"""
+                    lhs and rhs counts do not match.
+                    lhs count: \#(lhs.count)
+                    rhs count: \#(rhs.count)
+                    
+                    lhs: \#(lhs)
+                    
+                    rhs: \#(rhs)
+                    
+                    key path: \#(keyPathAsString(keyPath: keyPath))
+                """#, file: file, line: line)
+            }
+            return false
+        }
+        var finalResult = true
+        for (key, value) in lhs {
+            var keyPath = keyPath
+            keyPath.append(key)
+            finalResult = finalResult && assertEqual(lhs: value, rhs: rhs[key], keyPath: keyPath, file: file, line: line, shouldAssert: shouldAssert)
+        }
+        return finalResult
+    }
+    
+    /// Performs flexible testing assertions between two `[AnyCodable]` instances.
+    /// exactMatchTree = nil means no exact matching behavior
+    static func assertFlexibleEqual(validation: [String: AnyCodable]?, input: [String: AnyCodable]?, keyPath: [Any], exactMatchTree: [String: Any]?, file: StaticString = #file, line: UInt = #line, shouldAssert: Bool = true) -> Bool {
+        if validation == nil {
+            return true
+        }
+        guard let lhs = validation, let rhs = input else {
```

Comment:
Can we extract lhs, rhs non inwrapping and count of lhs and rhs extracted to a separate function?  I see it being repeated in our asserts.